docs/internal/arithmetic.md

```diff
@@ -0,0 +1,51 @@
+# Arithmetic
+
+Numeric types in Verona are nothing but classes with methods that implement known arithmetic functions.
+
+For example:
+```
+  let x : I8 & mut = 42;
```

Comment:
I forgot to mention in the meeting, but I don't think `I8 & mut` actually makes sense as a thing.  `I8 & imm` should be a valid type.  You can't modify the value `42`, you can only modify the variable `x` (if it were `var`, not `let`) to refer to a different integer.  

Conceptually, we have one concrete instance of every possible number of every numeric type.  As an optimisation, we store the actual number, rather than a pointer to an object containing the number, in the variable.  As a result of this optimisation, we can't actually express `I8 & mut` unless we add an allocation.

For a concrete illustration of the difference:

```
var y : I8 & mut = 42;
var z = y;
z++;
```

Do you expect the value of `y` to change because you've modified the object that `z` refers to?

Reply:
You're right. I'm still confused as to whether this is a special case of numeric types or literal assignment, but in this case, it doesn't matter. I can easily replace it with `I8 & imm` and not worry about it in this section.

But perhaps it would be interesting to follow that thread in a separate section of the same document.

In your example, `z++` is `constexpr` so directly evaluated to `43` a compile time, but:
```
  func(a: I8) {
    var y : I8 & mut = 42;
    var z = y;
    z += a;
  }
```
where `a` cannot be inferred at compile time. Here, `z` first has a "copy" of `42` and reassign whatever the sum is next.

Still, the storage space is the same: bit-width. Dynamic dispatch on `z.+=(a)` would take the descriptor of `z`'s type, call the offset of `+=` and pass `z` and `a`, storing back on the same place the result. (assuming we have the `+=` operator).

I think the same is true for strings:
```
  var a : CString = "abc";
  var b : CString = a;
  a.concat("def");
```
Here, also `constexpr`, `a = "abc"` and `b = "abcdef"`. Those would be references to compile-time constant literals, and if we uniquefy strings, equality can be just a pointer comparison. If the operations cannot be constant folded, then comparison is a `strcmp`.

Does that make sense?

Reply:
Strings are an entirely separate thing and it's still not clear to me what the correct behaviour for string literals should be (possibly similar to integers, where it can be unified with any string type during type inference?), but the string example is also similar and has nothing to do with constant propagation or constant expression evaluation:

```
var a : SomeStringType & mut = SomeStringType("foo");
var b = a; // b and a are now pointers to the same string object.
b.append("bar");
// Both a and b refer to the same object that now contains the string "foobar"
```

Whether we evaluate the `append` call at compile time or run time, both `a` and `b` in this example will point to the same string and that string will have been mutated.

You definitely don't want this behaviour for integers, because it would require every integer variable to actually be a pointer to an integer, rather than simply an integer.  This example:

```
var a : U32 & mut = 1;
var b = a;
a++;
```

Would look something like this in C:

```
int *a = malloc(sizeof(int));
*a = 1;
int &b = a;
(*a)++;
```

If the types are `imm`, the naïve C equivalent is:

```
int *a = malloc(sizeof(int));
*a = 1;
int &b = a;
int *tmp = malloc(sizeof(int));
*tmp = (*a) + 1;
a = tmp;
```

Because the value of the integer in the allocation never changes; however, we can perform a simple optimisation and embed the value in the pointer.  Because Verona knows that the underlying type is `U32`, rather than a generic pointer-to-some-kind-of-object, we can avoid needing any boxing tricks and just make the underlying type of the variable a 32-bit integer.

Reply:
That's why I used `CString` (to mean `const char`) but perhaps I should have said `CString & imm`. A generic string object is just like any other object, reference, mutability and all. I was just talking about constant literal sequences of chars.